src/defrag.c

```diff
@@ -689,11 +744,21 @@ void defragKey(defragCtx *ctx, dictEntry *de) {
             dictEntry *expire_de = kvstoreDictFindEntryByPtrAndHash(db->expires, slot, keysds, hash);
             if (expire_de) kvstoreDictSetKey(db->expires, slot, expire_de, newsds);
         }
+
+        /* Update the key's reference in the dict's metadata or the listpackEx. */
+        if (unlikely(ob->type == OBJ_HASH))
+            hashTypeUpdateKeyRef(ob, newsds);
     }
 
     /* Try to defrag robj and / or string value. */
-    ob = dictGetVal(de);
-    if ((newob = activeDefragStringOb(ob))) {
+    if (unlikely(ob->type == OBJ_HASH && hashTypeHasExpireField(ob))) {
+        /* Update its reference in the ebucket while defragging it. */
+        newob = ebDefragItem(&db->hexpires, &hashExpireBucketsType, ob, (ebDefragFunction *)activeDefragStringOb);
```

Comment:
The function should renamed to `hashTypeHasExpireMeta` or `hashTypeExpireSupport` . It doesn't necessarily gurantee that it has fields with expiry. Only tells that it support expiry. 

We need to verify either inside `ebDefragItem` or before the call that ExpireMeta of the hash is not `Trash`.

Reply:
I think we can discard entire function `hashTypeHasExpireMeta()` and use instead existing `hashTypeGetMinExpire()` against `EB_EXPIRE_TIME_INVALID`.